File: app/src/main/java/com/egecius/gtdx/ListActivityPresenterImpl.java
package com.egecius.gtdx;

import com.egecius.gtdx.datatypes.TodoTask;
import com.egecius.gtdx.db.Db;
import com.egecius.gtdx.db.DbImpl;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import rx.functions.Action1;

final class ListActivityPresenterImpl implements ListActivityPresenter {

	private Db db = new DbImpl(FirebaseDatabase.getInstance());
	private final ListActivityView listActivity;

	ListActivityPresenterImpl(final ListActivityView listActivity) {
		this.listActivity = listActivity;
	}

	@Override
	public void onCreate() {
		observeTasksInCloudDatabase();
	}

	private void observeTasksInCloudDatabase() {
		db.getAllTasks().subscribe(new Action1<Map<String, Map>>() {
			@Override
			public void call(final Map<String, Map> map) {
				updateTasks(map);
			}
		});
	}

	@Override
	public void onNewTaskAdded(final String taskTitle) {
		addTaskToDb(taskTitle);
	}

	private void addTaskToDb(final String taskTitle) {
		TodoTask taskToAdd = new TodoTask(taskTitle);
		db.addTask(taskToAdd);
	}

	private void updateTasks(final Map<String, Map> map) {
		List<TodoTask> list = extractToTaskList(map);
		listActivity.onTasksUpdated(list);
	}

	private List<TodoTask> extractToTaskList(final Map<String, Map> map) {

		List<TodoTask> list = new ArrayList<>();

		for (Map.Entry<String, Map> entry : map.entrySet()) {
			Map taskMap = entry.getValue();
			String title = (String) taskMap.get("title");
			TodoTask todoTask = new TodoTask(title);
			list.add(todoTask);
		}

		return list;
	}

}

File: app/src/main/kotlin/com/egecius/gtdx/ui/ListActivityPresenterImpl.kt
package com.egecius.gtdx.ui

import android.util.Log
import com.egecius.gtdx.datatypes.TodoTask
import com.egecius.gtdx.db.Db
import com.egecius.gtdx.utils.SorterImpl
import java.util.*

internal class ListActivityPresenterImpl(private val view: ListActivityView, val db: Db) : ListActivityPresenter {

    override fun onCreate() {
        observeTasksInCloudDatabase()
        view.showProgressBar()
    }

    private fun observeTasksInCloudDatabase() {
        db.getAllTasks.subscribe { map ->
            view.hideProgressBar()
            updateTasks(map) }
    }

    override fun onNewTaskAdded(taskTitle: String) {
        addTaskToDb(taskTitle)
    }

    override fun onRemoveTaskClicked(id: String) {
        db.removeTask(id)
    }

    private fun addTaskToDb(taskTitle: String) {
        db.addTask(taskTitle)
    }

    private fun updateTasks(map: Map<String, Map<*, *>>) {
        val unsortedList = extractToTaskList(map)
        val sortedList = sortByMostRecent(unsortedList)
        view.onTasksUpdated(sortedList)
    }

    private fun sortByMostRecent(unsortedList: List<TodoTask>): List<TodoTask> {
        return SorterImpl().sortNewestFirst(unsortedList)
    }

    private fun extractToTaskList(map: Map<String, Map<*, *>>): List<TodoTask> {

        val list = ArrayList<TodoTask>()

        for ((key, taskMap) in map) {
            val id = taskMap["id"] as String
            val title = taskMap["title"] as String
            val timestamp = taskMap["timestamp"] as Long
            val todoTask = TodoTask(id, title, timestamp)
            list.add(todoTask)
        }

        return list
    }

}
